### PR TITLE
Support for MultiRZ and CRot in default.qubit.tf

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**

Passing a TensorFlow `Variable` to the  `MultiRZ` operation with the `default.qubit.tf` device results in an error:
```python
import pennylane as qml
import tensorflow as tf
import numpy as np

x = tf.Variable(np.array([1.7]))

dev = qml.device('default.qubit.tf', wires=3)

@qml.qnode(dev, interface='tf')
def circuit(params):
    qml.MultiRZ(params[0], wires=[0])
    return qml.expval(qml.PauliZ(0))

circuit(x)
```
```
~/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py in convert_to_eager_tensor(value, ctx, dtype)
     96       dtype = dtypes.as_dtype(dtype).as_datatype_enum
     97   ctx.ensure_initialized()
---> 98   return ops.EagerTensor(value, ctx.device_name, dtype)
     99 
    100 

TypeError: Cannot convert (-0-1j) to EagerTensor of dtype double
```
propagated from:
```
~/xanadu/pennylane/pennylane/ops/qubit.py in _eigvals(cls, theta, n)
    836     @classmethod
    837     def _eigvals(cls, theta, n):
--> 838         return np.exp(-1j * theta / 2 * pauli_eigs(n))
```
This error arises because the `default.qubit.tf` device needs to handle internal data using TensorFlow data structures and operations. However, this device is using the definition of `MultiRZ` for `default.qubit` which does not use TF.

Custom supported operations are implemented in the [`tf_ops.py file`](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/devices/tf_ops.py) and wired in to the device using the [`parametric_ops dictionary`](https://github.com/PennyLaneAI/pennylane/blob/855327baf4e013d7c8eee43105c4680686eb75d3/pennylane/devices/default_qubit_tf.py#L132).

**Description of the Change:**
* Updates `default.qubit.tf` to support `MultiRZ`
* Temporarily removes `CRot` from the supported operations due to a similar error. `CRot` can be used, but the decomposition will be applied by default.

**Benefits:**
No errors when using `MultiRZ` and further PennyLane subpackages such as `qaoa` that depend on `MultiRZ`.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Thanks to @stfnmangini for discovering the issue!

### CODE REVIEW

File: tests/devices/test_default_qubit_tf.py [hunk 1]
```diff
@@ -72,9 +74,9 @@
 #####################################################
 
 single_qubit = [(qml.S, S), (qml.T, T), (qml.PauliX, X), (qml.PauliY, Y), (qml.PauliZ, Z), (qml.Hadamard, H)]
-single_qubit_param = [(qml.PhaseShift, Rphi), (qml.RX, Rotx), (qml.RY, Roty), (qml.RZ, Rotz)]
+single_qubit_param = [(qml.PhaseShift, Rphi), (qml.RX, Rotx), (qml.RY, Roty), (qml.RZ, Rotz), (qml.MultiRZ, MultiRZ1)]
 two_qubit = [(qml.CZ, CZ), (qml.CNOT, CNOT), (qml.SWAP, SWAP)]
-two_qubit_param = [(qml.CRX, CRotx), (qml.CRY, CRoty), (qml.CRZ, CRotz)]
+two_qubit_param = [(qml.CRX, CRotx), (qml.CRY, CRoty), (qml.CRZ, CRotz), (qml.MultiRZ, MultiRZ2)]
```
Comment (josh146) [inline tests/devices/test_default_qubit_tf.py:79]: :+1:

File: pennylane/devices/default_qubit_tf.py [hunk 2]
```diff
@@ -164,6 +165,12 @@ def __init__(self, wires, *, shots=1000, analytic=True):
         # prevent using special apply method for this gate due to slowdown in TF implementation
         del self._apply_ops["CZ"]
 
+        # TODO: add support for the CRot operation
+        # Remove CRot from the supported operations so that it will be
+        # decomposed by default
+        if "CRot" in self.operations:
+            self.operations.remove("CRot")
```
Comment (josh146) [inline pennylane/devices/default_qubit_tf.py:172]: How come `CRot` doesn't work? There appears to be a `CRot` function in `tf_ops`.

I think this can be better fixed just be including `CRot` in the `parametric_ops` dictionary 😆

File: pennylane/devices/tf_ops.py [hunk 3]
```diff
@@ -102,6 +103,21 @@ def Rot(a, b, c):
...
+    Args:
+        theta (float): rotation angle
+        n (int): Number of wires the rotation acts on
+
+    Returns:
+        tf.Tensor[complex]: diagonal part of the MultiRZ matrix
+    """
+    theta = tf.cast(theta, dtype=C_DTYPE)
+    multi_Z_rot_eigs = tf.exp(-1j * theta / 2 * pauli_eigs(n))
+    return tf.convert_to_tensor(multi_Z_rot_eigs)
```
Comment (josh146) [inline pennylane/devices/tf_ops.py:118]: nice!

File: pennylane/devices/tf_ops.py [hunk 4]
```diff
@@ -102,6 +103,21 @@ def Rot(a, b, c):
     return tf.linalg.diag(RZ(c)) @ RY(b) @ tf.linalg.diag(RZ(a))
 
 
+def MultiRZ(theta, n):
+    r"""Arbitrary multi Z rotation.
+
+    Args:
+        theta (float): rotation angle
+        n (int): Number of wires the rotation acts on
```
Comment (josh146) [inline pennylane/devices/tf_ops.py:111]:
```suggestion
        n (int): number of wires the rotation acts on
```

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks Antal! However, I think the CRot issue is due to it missing from the `parametric_ops` dictionary, as opposed to anything else.

Also, don't forget to update the 'bugfix' section of the changelog!

File: pennylane/devices/default_qubit_tf.py (hunk 2 above)
Comment (antalszava) [reply pennylane/devices/default_qubit_tf.py]: Yeah, thanks for spotting that! Wiring it in and testing that it passes when using a `TF.Variable` was missing. Added these in the latest commit.

File: tests/devices/test_default_qubit_tf.py [hunk 5]
```diff
@@ -926,6 +928,69 @@ def circuit():
...
+        dev = qml.device("default.qubit.tf", wires=1)
+        state = init_state(1)
+
+        @qml.qnode(dev, interface='tf')
+        def circuit(params):
+            qml.QubitStateVector(state, wires=[0])
+            op(params[0], wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        # Pass a TF Variable to the qfunc
```
Comment (antalszava) [inline tests/devices/test_default_qubit_tf.py:945]: Worth having tests like these along with the tests in `TestApply`, as the edge case that was missed involved integration with the QNode. Had one unit test for `CRot` previously that passed well.

**Comment by antalszava** [comment]
Updated with integration tests for `CRot` too. Thanks for the suggestions!

**Review by josh146 — APPROVED** [review]
Thanks @antalszava!
